### PR TITLE
Add concept sklearn nearest neighbors

### LINKED ISSUE #4895: [Concept Entry] Sklearn nearest-neighbours
### Reason/inspiration (optional)

We would like a new entry on the `Nearest Neighbors` concept under Sklearn. The entry should go in a new file under `docs/content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md`.

The entry should include:

- An introduction to the concept
- A `Syntax` section that provides the syntax for the concept (if applicable)
- An `Example` section that provides an example demonstrating the concept in use (if applicable)
- A `Codebyte Example` section that provides a codebyte example demonstrating the concept in use (if applicable)

Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by igarousi**
### Summary of change
Created brief documentation for nearest neighbors algorithms, focusing on the K-NN classification and regression models.

### Issue Solved
Closes #4895 (Sklearn nearest-neighbours)

### Type of Change
- Adding a new concept

### Checklist
- [ ] All writings are my own.  [_Please note that most of the content is sourced directly from the sklearn.neighbors` documentation and it is cited appropriately. The examples created are original. Technical and grammatical corrections were made using ChatGPT_].
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  - **Nearest Neighbors Regression** predicts the value of a new data point in cases where the data labels and values are continuous (not discrete) by (1) identifying the nearest neighbors of the new point from the training dataset based on a chosen distance metric, (2) averaging the values of these neighbors, and (3) assigning the new point the averaged value. For example, if the new point has 3 neighbors with values 10, 20, and 30, the new point's predicted value will be the average of these, i.e., 20. `scikit-learn` offers two different regressor: `KNeighborsRegressor` and `RadiusNeighborsRegressor`. Their learning are similar to those mentioned in the classification above. 
+
+
+## Codebyte Example (if applicable)
+
+Here are two simple examples for classification and regression using `KNeighborsClassifier` and `KNeighborsRegressor` from scikit-learn.
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:34]:
```suggestion
Here are two simple examples of classification and regression using `KNeighborsClassifier` and `KNeighborsRegressor` from scikit-learn.
```

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+### Usage of Nearest Neighbors in Scikit-learn for a **classification** problem:
+The following code uses the k-Nearest Neighbors (k-NN) classification algorithm to predict snow seasons (Accumulation or Melt) based on snow depth, air temperature, and snow density. The dataset contains historical records of these three features, alongside their corresponding seasons. The classification algorithm classifies new data points by finding the 5 nearest data points in the training set and taking a majority vote of their seasons. In this example, the search algorithm is 'auto', meaning the classifier will automatically select the most appropriate algorithm based on the training data provided. You can change it by specifying the `algorithm` parameter to other options when using `KNeighborsClassifier`. The method outputs a class label that is the result of a majority vote among the nearest neighbors, ensuring that the predicted class is one that exists in the training set (1 or 0 in this example) .
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:37]:
```suggestion
The following code uses the k-Nearest Neighbors (k-NN) classification algorithm to predict snow seasons (Accumulation or Melt) based on snow depth, air temperature, and snow density. The dataset contains historical records of these three features, alongside their corresponding seasons. The classification algorithm classifies new data points by finding the 5 nearest data points in the training set and taking a majority vote on their seasons. In this example, the search algorithm is 'auto', meaning the classifier will automatically select the most appropriate algorithm based on the training data provided. You can change it by specifying the `algorithm` parameter to other options when using `KNeighborsClassifier`. The method outputs a class label that is the result of a majority vote among the nearest neighbors, ensuring that the predicted class is one that exists in the training set (1 or 0 in this example).
```

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+```shell
+import numpy as np
+import pandas as pd
+from sklearn.neighbors import KNeighborsClassifier
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:42]:
```suggestion
```py
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
```
Code should always be written inside their respective code blocks depending on the language, here our code is written in Python language so we can use " `py` " here.
Shell block is used to show the output for example.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        36.4, 37.5, 38, 36.2, 36.2, 37.6, 38.2, 38.9, 39.7, 40.6,
+        43.1, 45.3, 45.7, 47.5, 47.2, 46.9, 45, 42.3, 45.1, 46.1,
+        47.1, 45.6, 45.7, 45.1, 44.9, 43.5, 46.1, 45.3, 45, 45],
+    "Season": [
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+}
+
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:67]:
```suggestion
# Create a DataFrame with sample data
data = {
    "Snow Depth (in)": [
        65, 64, 64, 69, 78, 75, 71, 70, 70, 68,
        66, 64, 64, 68, 68, 66, 65, 64, 63, 62,
        62, 59, 58, 55, 53, 51, 52, 56, 53, 51,
        48, 48, 46, 43, 41, 40, 36, 36, 36, 36],
    "Air Temperature Average (degF)": [
        33.4, 33.6, 16.5, 16.9, 23, 25, 25.5, 20.8, 18.9, 25,
        30.2, 31.6, 26.1, 23.2, 27.1, 28.6, 28.4, 29.5, 32, 32,
        31.5, 32.5, 40.8, 44.2, 43.7, 35.8, 34.3, 37.8, 33.4, 35.2,
        35.2, 38.7, 45.3, 41.4, 45, 42.6, 36.1, 32.7, 33.6, 38.3],
    "Snow Density (pct)": [
        32.6, 33.4, 34.5, 32.8, 30, 31.7, 33.5, 34, 34, 35,
        36.4, 37.5, 38, 36.2, 36.2, 37.6, 38.2, 38.9, 39.7, 40.6,
        43.1, 45.3, 45.7, 47.5, 47.2, 46.9, 45, 42.3, 45.1, 46.1,
        47.1, 45.6, 45.7, 45.1, 44.9, 43.5, 46.1, 45.3, 45, 45],
    "Season": [
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
}

```
This is totally correct, the data dictionary is correctly formatted.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+# Classification model
+knn_classifier = KNeighborsClassifier(n_neighbors=5)
+knn_classifier.fit(X, y)
+
+X_test = np.array([[29, 27, 45.5]])
+
+predictions = knn_classifier.predict(X_test)
+
+print(f'Predicted season for the given snow condition is: {["Accumulation" if predictions[0]==1 else "Melt"][0]}.')
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:83]:
```suggestion
print(f'Predicted season for the given snow condition is: {["Accumulation" if predictions[0]==1 else "Melt"][0]}.')
```
season="Accumulation" if predictions[0]==1 else "Melt"
print(f 'Predicted season for the given snow condition is: {season}.')
The use of these two above statements makes the code more readable and easier to understand, though it is not a necessary change.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+```
+
+### Usage of Nearest Neighbors in Scikit-learn for a **regression** problem:
+The following code predicts the discharge in cubic feet per second based on the snow water equivalent in inches using a k-Nearest Neighbors (k-NN) regression model. The snowmax is an array representing snow water equivalent measurements, and flowmax, is an array representing corresponding discharge values. The model is trained using these arrays. The parameter `n_neighbors=5` specifies that the prediction for a new data point will be based on the **average discharge values** of the 5 nearest neighbors in the training data. The search algorithm implicitly used by `KNeighborsRegressor` finds the nearest neighbors by calculating the distance between points in the feature space, typically using the Euclidean distance. The code then predicts the discharge for a new snow water equivalent value of 42 inches and prints the result. This method outputs a value that is the averages the output, leading to potentially new values not seen in the training set.
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:87]:
```suggestion
The following code predicts the discharge in cubic feet per second based on the snow water equivalent in inches using a k-Nearest Neighbors (k-NN) regression model. The snowmax is an array representing snow water equivalent measurements, and the flowmax is an array representing corresponding discharge values. The model is trained using these arrays. The parameter `n_neighbors=5` specifies that the prediction for a new data point will be based on the **average discharge values** of the 5 nearest neighbors in the training data. The search algorithm implicitly used by `KNeighborsRegressor` finds the nearest neighbors by calculating the distance between points in the feature space, typically using the Euclidean distance. The code then predicts the discharge for a new snow water equivalent value of 42 inches and prints the result. This method outputs a value that averages the output, which may include values that are not seen in the training set.
```

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+snow_test = np.array([42])
+
+# Regression model
+knn_regressor = KNeighborsRegressor(n_neighbors=5)
+knn_regressor.fit(snowmax.reshape(-1, 1), flowmax)
+predictions = knn_regressor.predict(snow_test.reshape(-1, 1))
+
+print(f'Predicted flow for snow water equivalent of {snow_test[0]} inches: {predictions[0]} cubic feet per second.')
+```
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:114]:
``` suggestion
```py
import numpy as np
from sklearn.neighbors import KNeighborsRegressor

# Snow water equivalent in inch
snowmax = np.array([44.3, 43.6, 21.1, 58.4, 48.5, 51. , 38.7, 53.3, 16.8, 27. , 46.4,
       24.6, 29.7, 16.9, 41.2, 28.7, 34.9, 42.7, 55.9, 45. , 49.6, 34.7,
       25.4, 29.8, 34.5, 31.5, 45.2, 58.2, 26.1, 42. , 42.1, 28.9, 66.2,
       29.7, 24.2, 48.5, 27.6, 37.6, 54.2, 32.3, 41.5, 41.7, 26.4])  

# Discharge in cubic feet per second
flowmax = np.array([ 855.,  876., 584., 1270., 1400., 1830., 865., 1870.,  338.,
        540.,  748., 290.,  760., 243., 1180., 595.,  1220., 1330.,
        1560., 977., 1430., 562., 721., 492.,  716.,  529., 1230.,
        1530., 468., 722.,  967., 852., 1630., 429.,  480., 773.,
        545.,  775., 1390., 628., 929., 695.,  271.])

snow_test = np.array([42])

# Regression model
knn_regressor = KNeighborsRegressor(n_neighbors=5)
knn_regressor.fit(snowmax.reshape(-1, 1), flowmax)
predictions = knn_regressor.predict(snow_test.reshape(-1, 1))

print(f'Predicted flow for snow water equivalent of {snow_test[0]} inches: {predictions[0]} cubic feet per second.')
```
```
The code should always be written inside their supportive code blocks but not inside the shell block. For more information check this [documentation](documentation/style-guide.md)

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hello Author,
I reviewed your file and found some corrections to be made before this PR merges. Kindly have a look at my suggestions and try to rewrite it as soon as possible and let me know.
Thanks

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md (hunk 4 above)
Comment (igarousi) [reply content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:67]: It seems no change is required for this block of code.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md (hunk 5 above)
Comment (igarousi) [reply content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:83]: Good suggestion!

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md (hunk 7 above)
Comment (igarousi) [reply content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:114]: Done!

**Comment by igarousi** [comment]
Thanks for your review. All suggestions have been applied.

**Review by SaviDahegaonkar — APPROVED** [review]

**Comment by SaviDahegaonkar** [comment]
Hey @igarousi,
I had a look at the changes you made on the review and approved them.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  - 'Scikit-learn'
+  - 'K-Nearest Neighbors'
+  - 'Machine Learning'
+  - 'Classification'
+  - 'Regression'
+CatalogContent: 
+  - 'learn-python-3'
+  - 'paths/computer-science'
+  - 'paths/data-science'
+  - 'paths/data-science-foundations'
```
Comment (mamtawardhani) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:18]: Please shorten the description, as it cannot go beyond 160 characters. Please remove the tags 'classification' and 'regression', and in the catalog content, keep only the first two bullet points.

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+---
+
+**Nearest Neighbors Algorithms** are non-parametric, neighbor-based learning methods that predict labels and values by using the similarity between data points. This fundamental concept is at the core of many learning routines and is especially useful in scenarios where the relationships between data points are more important than their distribution. The *number of points* or neighbors used during the training process can be specified by the user (as in the k-nearest neighbor learning method) or adjusted based on the local density of points (as in radius-based neighbor learning). The *distance* metric can vary, with the standard Euclidean distance being a common choice. Nearest neighbors algorithms are applied in classification, regression, and clustering problems. They are widely used due to their simplicity in various real-world applications; however, they can become computationally intensive for large datasets. Please note that most of the content on this page is sourced directly from the `sklearn.neighbors` [documentation](https://scikit-learn.org/stable/modules/neighbors.html#neighbors).
```
Comment (mamtawardhani) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:21]: Could you shorten the explanation here?

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+  - **Nearest Neighbors Regression** predicts the value of a new data point in cases where the data labels and values are continuous (not discrete) by (1) identifying the nearest neighbors of the new point from the training dataset based on a chosen distance metric, (2) averaging the values of these neighbors, and (3) assigning the new point the averaged value. For example, if the new point has 3 neighbors with values 10, 20, and 30, the new point's predicted value will be the average of these, i.e., 20.0. The `scikit-learn` package offers two different regressors: `KNeighborsRegressor` and `RadiusNeighborsRegressor`. Their learning methods are similar to those mentioned in the classification above. 
+
```
Comment (mamtawardhani) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:30]: Same for this, could you please shorten the applications, let us follow the "Addressing more in less words" concept here. The current entry is having too much text

File: content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+## Codebyte Example 
+
+Here are two simple examples of classification and regression using `KNeighborsClassifier` and `KNeighborsRegressor` from scikit-learn.
```
Comment (mamtawardhani) [inline content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:34]: This is incorrect, the entry is on the `NearestNeighbors` term. This is where we can import it from:
```
from sklearn.neighbors import NearestNeighbors
```

Comment (mamtawardhani) [reply content/sklearn/concepts/nearest-neighbors/nearest-neighbors.md:34]: You will have to change the codes for this entry since the issue is for the from `NearestNeighbors`, the Classifier and Regressor will be other term entries under this.

**Review by mamtawardhani — CHANGES_REQUESTED** [review]
Hey @igarousi, the entry will require some changes:
1. Please change the meta-data as per the comment
2. Please reduce the content, it is not following the template. The entry has too much text.
3. The codes should be for the `NearestNeighbors` term, so the ##Example and ##codebyte should be based on that.

Let me know once you make the changes. Than you! 😃

**Comment by mamtawardhani** [comment]
@igarousi are you still working on this? If you don't make the changes within the next 5 days, I'll have to close this PR.

**Comment by mamtawardhani** [comment]
Closing this PR due to the inactivity of the author. @igarousi
